### PR TITLE
feature: Support choosing Kubernetes QoS class through the decorator

### PR DESCRIPTION
**Pull request by saikonen**
enables choosing the Kubernetes QoS class out of `Guaranteed/Burstable/BestEffort` through the decorator.

### CODE REVIEW

File: metaflow/plugins/kubernetes/kubernetes_decorator.py [hunk 1]
```diff
@@ -109,6 +111,8 @@ class KubernetesDecorator(StepDecorator):
     hostname_resolution_timeout: int, default 10 * 60
         Timeout in seconds for the workers tasks in the gang scheduled cluster to resolve the hostname of control task.
         Only applicable when @parallel is used.
+    qos_class: str, optional, default: None
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_decorator.py:114]: defaults to `Burstable`?

File: metaflow/plugins/kubernetes/kube_utils.py [hunk 2]
```diff
@@ -23,3 +23,26 @@ def parse_cli_options(flow_name, run_id, user, my_runs, echo):
...
+def qos_requests_and_limits(qos_class: str, cpu: int, memory: int):
+    "return resource requests and limits for the kubernetes pod based on the given QoS Class"
+    # Determine the requests and limits to define chosen QoS class
+    qos_limits = {}
+    qos_requests = {}
+    if qos_class == "Guaranteed":
+        # Guaranteed - has both cpu/memory limits. requests not required, as these will be inferred.
+        qos_limits = {
+            "cpu": str(cpu),
+            "memory": "%sM" % str(memory),
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kube_utils.py:37]: what about storage?

File: metaflow/plugins/kubernetes/kubernetes_decorator.py (hunk 1 above)
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kubernetes_decorator.py:114]: might be useful to override the default through config as well

File: metaflow/plugins/kubernetes/kube_utils.py (hunk 2 above)
Comment (savingoyal) [reply metaflow/plugins/kubernetes/kube_utils.py:37]: just ephemeral-storage. gpu's always need to specified with both requests and limits. also, can you verify airflow too?

File: metaflow/plugins/kubernetes/kube_utils.py [hunk 3]
```diff
@@ -23,3 +23,26 @@ def parse_cli_options(flow_name, run_id, user, my_runs, echo):
...
 
     return flow_name, run_id, user
+
+
+def qos_requests_and_limits(qos: str, cpu: int, memory: int):
+    "return resource requests and limits for the kubernetes pod based on the given QoS Class"
+    # Determine the requests and limits to define chosen QoS class
+    qos_limits = {}
+    qos_requests = {}
+    if qos == "Guaranteed":
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kube_utils.py:33]: maybe a case insensitive check?

Comment (saikonen) [reply metaflow/plugins/kubernetes/kube_utils.py:33]: there's validation in the kubernetes decorator that is case-sensitive as well. I can also make all defaults lower case but `besteffort` reads a bit clunky.

File: metaflow/plugins/kubernetes/kube_utils.py (hunk 2 above)
Comment (saikonen) [reply metaflow/plugins/kubernetes/kube_utils.py:37]: handling storage as well now. For the `BestEffort` case, I figure we keep the ephemeral-storage in the requests, or do we omit it completely?

File: metaflow/plugins/kubernetes/kube_utils.py [hunk 4]
```diff
@@ -27,17 +27,19 @@ def parse_cli_options(flow_name, run_id, user, my_runs, echo):
...
         # Guaranteed - has both cpu/memory limits. requests not required, as these will be inferred.
         qos_limits = {
             "cpu": str(cpu),
             "memory": "%sM" % str(memory),
             "ephemeral-storage": "%sM" % str(storage),
         }
-    elif qos == "BestEffort":
+    elif qos == "besteffort":
         # BestEffort - no limit or requests for cpu/memory
         qos_requests = {"ephemeral-storage": "%sM" % str(storage)}
```
Comment (savingoyal) [inline metaflow/plugins/kubernetes/kube_utils.py:44]: should we even support best effort at the moment? there doesn't seem to be a use case for it at the moment

File: metaflow/plugins/kubernetes/kube_utils.py (hunk 3 above)
Comment (saikonen) [reply metaflow/plugins/kubernetes/kube_utils.py]: validation changed to case insensitive.

**Comment by savingoyal** [comment]
lgtm! can you verify airflow works as expected?

**Pull request merged**